jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -804,32 +803,78 @@ public void execute(Runnable job)
             // a job from the queue are not counted.
             int idle = AtomicBiInteger.getLo(counts);
 
-            // Start a thread if we have insufficient idle threads to meet demand
+            // Start a thread if we have not enough idle threads to meet demand,
             // and we are not at max threads.
-            startThread = (idle <= 0 && threads < _maxThreads) ? 1 : 0;
+            startThread = (idle <= 0 && threads < _maxThreads);
+
+            // If the job queue is bounded and we are at capacity, then we reject the job
+            if (_capacity >= 0 && idle <= 0 && !startThread && (threads - idle) >=  (_maxThreads + _capacity))
+                throw new RejectedExecutionException(job.toString());
 
             // Add 1|0 or 0|-1 to counts depending upon the decision to start a thread or not;
-            // idle can become negative which means there are queued tasks.
-            if (!_counts.compareAndSet(counts, threads + startThread, idle + startThread - 1))
-                continue;
+            // idle can become negative, which means there are queued tasks.
+            if (_counts.compareAndSet(counts, threads + (startThread ? 1 : 0), idle + (startThread ? 0 : -1)))
+                break;
+        }
 
-            break;
+        // We should always be able to add the job to the queue as we checked the capacity above,
+        // However, a starting thread may not yet have polled the queue, so this could fail in a race.
+        // Return if we are able to queue the job
+        if (_jobs.offer(job))
+        {
+            // Start additional threads
+            if (startThread)
+                startThread();
+            return;
         }
 
-        if (!_jobs.offer(job))
+        // Start additional threads
+        if (startThread)
+            startThread();
+
+        if (_putSupported)
         {
-            // reverse our changes to _counts.
-            if (addCounts(-startThread, 1 - startThread))
-                LOG.warn("{} rejected {}", this, job);
-            throw new RejectedExecutionException(job.toString());
+            // We lost the race with a starting thread, so we can do a blocking put that will wait for the starting thread
+            try
+            {
+                _jobs.put(job);
+                return;
+            }
+            catch (InterruptedException e)
+            {
+                addCounts(0, 1);
+                Thread.currentThread().interrupt();
+                throw new RejectedExecutionException(e);
+            }
+            catch (UnsupportedOperationException uoe)
+            {
+                LOG.warn("Detected bounded thread pool queue {} without put() support", _jobs.getClass());
```

Comment:
If `put()` is not implemented, we will get this WARN a million times in the logs.

Reply:
No, only once because of the boolean below

Reply:
No it was protected from that.... but removed now anyway